daemon/smartcard.c

```diff
@@ -177,6 +177,34 @@ smartcard_send_recv_block(const DaemonToToken *out)
 	return msg;
 }
 
+bool
+smartcard_cert_has_valid_format(unsigned char *cert_buf, size_t cert_buf_len)
+{
+	const char *begin_cert_str = "-----BEGIN CERTIFICATE-----\n";
+	const char *end_cert_str = "-----END CERTIFICATE-----\n";
+	size_t begin_cert_str_len = strlen(begin_cert_str);
+	size_t end_cert_str_len = strlen(end_cert_str);
+
+	if (cert_buf == NULL || cert_buf_len == 0) {
+		ERROR("Given certificate is empty.");
+		return false;
+	}
+	if (cert_buf_len < end_cert_str_len + begin_cert_str_len) {
+		ERROR("Invalid certificate length %zu.", cert_buf_len);
+		return false;
+	}
+	if (memcmp(cert_buf, begin_cert_str, begin_cert_str_len)) {
+		ERROR("Invalid certificate: begin string not found.");
+		return false;
+	}
+	if (memcmp((const char *)cert_buf + (cert_buf_len - end_cert_str_len), end_cert_str,
```

Comment:
as memcmp expects a const void * i'd suggest to cast to const void*

Reply:
I rewrote that line of code to 
`if (memcmp(cert_buf + sizeof(char) * (cert_buf_len - end_cert_str_len), end_cert_str,  ...`
for better legibility.